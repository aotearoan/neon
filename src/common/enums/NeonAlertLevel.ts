/**
 * Alert levels are the severity levels to be used with <a href="/feedback/alert#examples">NeonAlert & NeonToast</a>
 * components to inform the user of the severity of a message.
 */
export enum NeonAlertLevel {
  /** Information level, use for standard informational messages. */
  Info = 'info',
  /** Success level, use for indicating successful actions. */
  Success = 'success',
  /** Warning level, use for indicating a problem that the user <strong>can</strong> recover from. */
  Warn = 'warn',
  /** Error level, use for indicating a problem that the user <strong>cannot</strong> recover from. */
  Error = 'error',
}
