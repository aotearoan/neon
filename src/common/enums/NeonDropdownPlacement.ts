/**
 * Describes placement of dropdown related components like:
 * <ul class="no-style">
 *  <li><a href="/user-input/select">NeonSelect</a></li>
 *  <li><a href="/navigation/dropdown-menu">NeonDropdownMenu</a></li>
 *  <li><a href="/presentation/dropdown">NeonDropdown</a></li>
 * </ul>
 *
 * <p>The placement has a <strong>primary placement</strong> (the first direction) & a <strong>secondary placement</strong>
 * (the second direction). The primary placement is the location of the popup relative to the triggering element (e.g.
 * the select button). With the secondary placement defining the alignment relative to the trigger. E.g. BottomLeft will
 * open the dropdown below the trigger element with the left-hand edges aligned.</p>
 *
 * <p>The placement utilities are smart & dynamic. They listen to resize events & aim for best effort based on where the
 * dropdown is opened. E.g. in the case above with BottomLeft if there is not enough visible room to align the component
 * to the left then it will try to align to the right & likewise if there is not enough space below the trigger it will
 * align to the top (& reverse the option order in the case of a select).</p>
 */
export enum NeonDropdownPlacement {
  /** Placement above trigger element, aligned to the left */
  TopLeft = 'top-left',
  /** Placement above trigger element, aligned to the right */
  TopRight = 'top-right',
  /** Placement below trigger element, aligned to the left */
  BottomLeft = 'bottom-left',
  /** Placement below trigger element, aligned to the right */
  BottomRight = 'bottom-right',
  /** Placement to the left of the trigger element, aligned to the top */
  LeftTop = 'left-top',
  /** Placement to the left of the trigger element, aligned to the bottom */
  LeftBottom = 'left-bottom',
  /** Placement to the right of the trigger element, aligned to the top */
  RightTop = 'right-top',
  /** Placement to the right of the trigger element, aligned to the bottom */
  RightBottom = 'right-bottom',
}
