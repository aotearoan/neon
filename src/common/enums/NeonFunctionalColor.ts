/**
 * The core color palettes available for styling components in Neon. See the design section
 * <a href="/design/color">Color</a> page for more details on color palettes.
 */
export enum NeonFunctionalColor {
  /** Low contrast is a more subtle neutral palette which can be used for secondary actions & supporting text. */
  LowContrast = 'low-contrast',
  /** Neutral is the base greyscale palette. */
  Neutral = 'neutral',
  /**
   * High contrast is a higher contrast palette than can be used for increased accessibility or for a stronger black &
   * white application.
   * */
  HighContrast = 'high-contrast',
  /** Brand is the main brand color palette, this is useful for branded CTAs, text & controls. */
  Brand = 'brand',
  /** Primary is the main application color palette. For most applications this will be used by default. */
  Primary = 'primary',
  /** Info can be used to convey informational context, e.g. notes & alerts */
  Info = 'info',
  /** Success can be used to convey successful actions. */
  Success = 'success',
  /** Warn can be used to convey issues/actions that are less severe but may still impact the user. */
  Warn = 'warn',
  /** Error can be used to convey issues/actions that are severe & that the user may not be able to recover from. */
  Error = 'error',
}
