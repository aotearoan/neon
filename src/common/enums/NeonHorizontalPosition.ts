/**
 * Describes the horizontal positioning of certain elements in the context of a component. An example would be an icon
 * in a button or label as well as the label for a checkbox or switch.
 */
export enum NeonHorizontalPosition {
  /** Left horizontal positioning. */
  Left = 'left',
  /** Right horizontal positioning. */
  Right = 'right',
}
