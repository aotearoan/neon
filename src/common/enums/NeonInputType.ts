/**
 * Corresponds to the <a href="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#input_types">HTML input
 * types</a> for use with <a href="/user-input/input">NeonInput</a>.
 *
 * NOTE: prefer the specific components for Number, Password, Date, File, Search fields.
 */
export enum NeonInputType {
  /** Date input for selecting dates from a calendar. Use <a href="/user-input/date-picker">NeonDatePicker</a>. */
  Date = 'date',
  /** Date time input for selecting dates including a specific time. */
  DateTime = 'datetime-local',
  /** Email input for entering email addresses. */
  Email = 'email',
  /**
   * File input for opening a file dialog. Use <a href="/user-input/file">NeonDatePicker</a> or
   * <a href="/user-input/drop-zone">NeonDropZone</a>.
   */
  File = 'file',
  /** Number input for entering numeric values. Use <a href="/user-input/number">NeonNumber</a> */
  Number = 'number',
  /** Password input for entering passwords. Use <a href="/user-input/password">NeonPassword</a> */
  Password = 'password',
  /**
   * Range input for a slider control. Use <a href="/user-input/slider">NeonSlider</a> or
   * <a href="/user-input/range-slider">NeonRangeSlider</a>
   */
  Range = 'range',
  /** Search input for searching. Use <a href="/user-input/search">NeonSearch</a> */
  Search = 'search',
  /** Telephone number input. */
  Tel = 'tel',
  /** Text input for single or multiline text input. */
  Text = 'text',
  /** Time input. */
  Time = 'time',
  /** URL input for entering web addresses. */
  Url = 'url',
}
