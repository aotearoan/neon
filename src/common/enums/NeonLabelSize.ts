/**
 * The size options for the <a href="/presentation/label">NeonLabel</a> component.
 */
export enum NeonLabelSize {
  /** Extra extra small, default is 16 rem. */
  ExtraExtraSmall = 'xxs',
  /** Extra small, default is 20 rem. */
  ExtraSmall = 'xs',
  /** Small, default is 28 rem. */
  Small = 's',
  /** Medium, default is 36 rem. */
  Medium = 'm',
  /** Large, default is 44 rem. */
  Large = 'l',
}
