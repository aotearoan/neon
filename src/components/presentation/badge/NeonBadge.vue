<template>
  <div
    :class="[
      `neon-badge--${color}`,
      alternateColor ? `neon-badge--alternate-color-${alternateColor}` : '',
      `neon-badge--${size}`,
      {
        'neon-badge--with-label': label,
        'neon-badge--disabled': disabled,
        'neon-badge--with-image': image,
        'neon-badge--with-icon': icon,
      },
      circular ? 'neon-badge--circular' : 'neon-badge--square',
    ]"
    class="neon-badge"
  >
    <div v-if="jazziconId" class="neon-badge__image" v-html="svg" />
    <span v-else-if="label" class="neon-badge__label">{{ label }}</span>
    <img v-else-if="image" :alt="imageAlt" :src="image" class="neon-badge__image" />
    <neon-icon v-else-if="icon" :disabled="disabled" :inverse="!!color" :name="icon" class="neon-badge__icon" />
  </div>
</template>

<script lang="ts" src="./NeonBadge.ts" />
