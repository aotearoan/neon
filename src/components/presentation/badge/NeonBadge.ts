import { computed, defineComponent } from 'vue';
import { NeonSize } from '@/common/enums/NeonSize';
import { NeonFunctionalColor } from '@/common/enums/NeonFunctionalColor';
import NeonIcon from '@/components/presentation/icon/NeonIcon.vue';
import { NeonJazziconUtils } from '@/common/utils/NeonJazziconUtils';
import { NeonColorUtils } from '@/common/utils/NeonColorUtils';

/**
 * A badge is a small square or circular component for representing user avatars. These can be in the form of an image, an icon or a two character string (e.g. the user's initials).
 */
export default defineComponent({
  name: 'NeonBadge',
  components: {
    NeonIcon,
  },
  props: {
    /**
     * The two character <em>initials</em> to display on the badge.
     */
    label: { type: String, default: null },
    /**
     * URL of the image to display on the badge.
     */
    image: { type: String, default: null },
    /**
     * An icon to display on the badge.
     */
    icon: { type: String, default: null },
    /**
     * If true, render the badge as a circle, instead of a square.
     */
    circular: { type: Boolean, default: false },
    /**
     * The size of the badge - Small, Medium or Large.
     */
    size: { type: String as () => NeonSize, default: NeonSize.Medium },
    /**
     * The color of the badge. This is one of the provided NeonFunctionalColors.
     */
    color: { type: String as () => NeonFunctionalColor, default: NeonFunctionalColor.LowContrast },
    /**
     * Alternate color for creating gradient badges. NOTE: can also be the same color as 'color'.
     */
    alternateColor: { type: String as () => NeonFunctionalColor, default: null },
    /**
     * Display the badge in the disable style
     */
    disabled: { type: Boolean, default: false },
    /**
     * Apply the generated Jazzicon style based on the unique identified provided (e.g. a wallet address, name, etc)
     */
    jazziconId: { type: String, default: null },
    /**
     * Badge image alt text.
     */
    imageAlt: { type: String, default: 'Badge' },
  },
  setup(props) {
    const getColor = (key: NeonFunctionalColor) => {
      let colorString = getComputedStyle(document.documentElement).getPropertyValue(`--neon-rgb-${key}-l1`);
      if (colorString.length === 0) {
        colorString = '0, 0, 0';
      }
      const colorRgb = colorString
        .trim()
        .split(', ')
        .map((str) => +str);
      return NeonColorUtils.rgbToHex(colorRgb);
    };

    const palette = computed(() => [
      '#000000',
      getColor(NeonFunctionalColor.Brand),
      getColor(NeonFunctionalColor.Primary),
      getColor(NeonFunctionalColor.Info),
      getColor(NeonFunctionalColor.Success),
      getColor(NeonFunctionalColor.Warn),
      getColor(NeonFunctionalColor.Error),
    ]);

    const svg = computed(() =>
      props.jazziconId
        ? NeonJazziconUtils.genSvg(
          palette.value,
          props.jazziconId,
          props.size === NeonSize.Small ? 32 : props.size === NeonSize.Medium ? 40 : 48,
        )
        : null,
    );

    return {
      svg,
    };
  },
});
