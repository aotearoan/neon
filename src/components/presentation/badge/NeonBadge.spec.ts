import NeonBadge from './NeonBadge.vue';
import { NeonSize } from '@/common/enums/NeonSize';
import { NeonFunctionalColor } from '@/common/enums/NeonFunctionalColor';
import { render } from '@testing-library/vue';

describe('NeonBadge', () => {
  it('renders label', () => {
    const label = 'test label';
    const { container } = render(NeonBadge, { props: { label } });
    expect(container.getElementsByClassName('neon-badge--with-label')[0]).toBeDefined();
    expect(container.getElementsByClassName('neon-badge__label')[0].textContent).toEqual(label);
  });

  it('renders image', () => {
    // given
    const image = '/test.jpg';
    const { container } = render(NeonBadge, { props: { image } });
    expect(container.getElementsByClassName('neon-badge--with-image')[0]).toBeDefined();
    const img = container.getElementsByClassName('neon-badge__image')[0] as HTMLImageElement;
    expect(img.getAttribute('src')).toEqual(image);
    expect(img.getAttribute('alt')).toEqual('Badge');
  });

  it('renders jazzicon', () => {
    // given
    const { html } = render(NeonBadge, { props: { jazziconId: 'Test Jazzicon' } });

    expect(html()).toMatchSnapshot();
  });

  it('renders image with imageAlt', () => {
    // given
    const image = '/test.jpg';
    const { container } = render(NeonBadge, { props: { image, imageAlt: 'xdd' } });
    // when
    const img = container.getElementsByClassName('neon-badge__image')[0] as HTMLImageElement;
    // then
    expect(img.getAttribute('src')).toEqual(image);
    expect(img.getAttribute('alt')).toEqual('xdd');
  });

  it('renders icon', () => {
    // given
    const icon = 'check';
    const { container } = render(NeonBadge, { props: { icon } });
    expect(container.getElementsByClassName('neon-badge--with-icon')[0]).toBeDefined();
    expect(container.getElementsByClassName('neon-badge__icon')[0]).toBeDefined();
  });

  it('renders square class', () => {
    const icon = 'check';
    const { container } = render(NeonBadge, { props: { icon } });
    expect(container.getElementsByClassName('neon-badge--square')[0]).toBeDefined();
  });

  it('renders circular class', () => {
    const icon = 'check';
    const { container } = render(NeonBadge, { props: { icon, circular: true } });
    expect(container.getElementsByClassName('neon-badge--circular')[0]).toBeDefined();
  });

  it('renders default size', () => {
    const icon = 'check';
    const { container } = render(NeonBadge, { props: { icon } });
    expect(container.getElementsByClassName('neon-badge--m')[0]).toBeDefined();
  });

  it('renders size', () => {
    const icon = 'check';
    const { container } = render(NeonBadge, { props: { icon, size: NeonSize.Small } });
    expect(container.getElementsByClassName('neon-badge--s')[0]).toBeDefined();
  });

  it('renders default color', () => {
    const icon = 'check';
    const { container } = render(NeonBadge, { props: { icon } });
    expect(container.getElementsByClassName('neon-badge--low-contrast')[0]).toBeDefined();
  });

  it('renders color', () => {
    const icon = 'check';
    const { container } = render(NeonBadge, { props: { icon, color: NeonFunctionalColor.Primary } });
    expect(container.getElementsByClassName('neon-badge--primary')[0]).toBeDefined();
  });

  it('renders alternateColor', () => {
    const icon = 'check';
    const { container } = render(NeonBadge, { props: { icon, alternateColor: NeonFunctionalColor.Primary } });
    expect(container.getElementsByClassName('neon-badge--alternate-color-primary')[0]).toBeDefined();
  });

  it('renders disabled', () => {
    const icon = 'check';
    const { container } = render(NeonBadge, { props: { icon, disabled: true } });
    expect(container.getElementsByClassName('neon-badge--disabled')[0]).toBeDefined();
  });
});
