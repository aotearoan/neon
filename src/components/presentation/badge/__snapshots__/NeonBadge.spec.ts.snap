// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NeonBadge renders jazzicon 1`] = `
<div class="neon-badge--low-contrast neon-badge--m neon-badge--square neon-badge">
  <div class="neon-badge__image"><svg xmlns="http://www.w3.org/2000/svg" x="0" y="0" width="40" height="40">
      <rect x="0" y="0" width="40" height="40" transform="translate(0 0) rotate(0.0 20 20)" fill="#000000"></rect>
      <rect x="0" y="0" width="40" height="40" transform="translate(2.650148400696 8.247754092555605) rotate(190.1 20 20)" fill="#000000"></rect>
      <rect x="0" y="0" width="40" height="40" transform="translate(-14.763630954718767 9.149803696525677) rotate(248.4 20 20)" fill="#000000"></rect>
      <rect x="0" y="0" width="40" height="40" transform="translate(25.82597178845803 -20.278135901172035) rotate(402.3 20 20)" fill="#000000"></rect>
    </svg></div>
</div>
`;
